packages/@glimmer/reference/lib/validators.ts

```diff
@@ -1,398 +1,331 @@
+  return tag[VALIDATE](snapshot);
 }
 
-export class DirtyableTag extends RevisionTag {
-  static create(revision = $REVISION) {
-    return new TagWrapper(this.id, new DirtyableTag(revision));
-  }
+//////////
 
-  private revision: Revision;
+/**
+ * This enum represents all of the possible tag types for the monomorphic tag class.
+ * Other custom tag classes can exist, such as CurrentTag and VolatileTag, but for
+ * performance reasons, any type of tag that is meant to be used frequently should
+ * be added to the monomorphic tag.
+ */
+const enum MonomorphicTagTypes {
+  Dirtyable,
+  Updatable,
+  Combinator,
+  Constant,
+}
 
-  constructor(revision = $REVISION) {
-    super();
-    this.revision = revision;
-  }
+const DIRTY: unique symbol = symbol('TAG_DIRTY');
+const UPDATE: unique symbol = symbol('TAG_UPDATE');
 
-  value(): Revision {
-    return this.revision;
-  }
+const TYPE: unique symbol = symbol('TAG_TYPE');
+const UPDATE_SUBTAGS: unique symbol = symbol('TAG_UPDATE_SUBTAGS');
 
-  dirty() {
-    this.revision = ++$REVISION;
-  }
+interface MonomorphicTagBase<T extends MonomorphicTagTypes> extends Tag {
+  [TYPE]: T;
 }
 
-register(DirtyableTag);
+export interface DirtyableTag extends MonomorphicTagBase<MonomorphicTagTypes.Dirtyable> {
+  [DIRTY](): void;
+}
 
-export function combineTagged(tagged: ReadonlyArray<Tagged>): Tag {
-  let optimized: Tag[] = [];
+export interface UpdatableTag extends MonomorphicTagBase<MonomorphicTagTypes.Updatable> {
+  [DIRTY](): void;
+  [UPDATE](tag: Tag): void;
+}
 
-  for (let i = 0, l = tagged.length; i < l; i++) {
-    let tag = tagged[i].tag;
-    if (tag === VOLATILE_TAG) return VOLATILE_TAG;
-    if (tag === CONSTANT_TAG) continue;
-    optimized.push(tag);
-  }
+export interface CombinatorTag extends MonomorphicTagBase<MonomorphicTagTypes.Combinator> {}
+export interface ConstantTag extends MonomorphicTagBase<MonomorphicTagTypes.Constant> {}
 
-  return _combine(optimized);
+interface MonomorphicTagMapping {
+  [MonomorphicTagTypes.Dirtyable]: DirtyableTag;
+  [MonomorphicTagTypes.Updatable]: UpdatableTag;
+  [MonomorphicTagTypes.Combinator]: CombinatorTag;
+  [MonomorphicTagTypes.Constant]: ConstantTag;
 }
 
-export function combineSlice(slice: Slice<Tagged & LinkedListNode>): Tag {
-  let optimized: Tag[] = [];
+type MonomorphicTag = UnionToIntersection<MonomorphicTagMapping[MonomorphicTagTypes]>;
+type MonomorphicTagType = UnionToIntersection<MonomorphicTagTypes>;
 
-  let node = slice.head();
+export class MonomorphicTagImpl implements MonomorphicTag {
+  private revision: Revision = INITIAL;
+  protected lastChecked: Revision = INITIAL;
+  protected lastValue: Revision = INITIAL;
 
-  while (node !== null) {
-    let tag = node.tag;
+  private isUpdating = false;
+  private subtag: Tag | null = null;
+  private subtags: Tag[] | null = null;
 
-    if (tag === VOLATILE_TAG) return VOLATILE_TAG;
-    if (tag !== CONSTANT_TAG) optimized.push(tag);
+  [TYPE]: MonomorphicTagType;
```

Comment:
I may just be not understanding what the actual `MonomorphicTag` type is, but why do you have to use the brands here instead of the const enum?

Reply:
It's because `TYPE` is the intersection of all of the types on all of the classes that make up the `MonomorphicTag` type (it being the intersection of all the subtypes), which is slightly different than the actual enum. This is part of what gets us the ability to magically do `new MonomorhpicTagImpl(TagType.Dirtyable)` and have that type to the `DirtyableTag` interface.

@krisselden may be able to explain better, this was his dark type magic that I'm still trying to wrap my head around 😄